soc/xtensa/intel_adsp/cavs/power.c

```diff
@@ -107,7 +107,6 @@ void pm_state_set(enum pm_state state, uint8_t substate_id)
 			/* do power down - this function won't return */
 			power_down_cavs(true, uncache_to_cache(&hpsram_mask[0]));
 		} else {
-			z_xt_ints_on(core_desc[cpu].intenable);
```

Comment:
This seems suspicious?  How does the CPU wake up on e.g. a timer interrupt then?  The k_cpu_idle() API is expressly intended to be a "wait for an interrupt" mechanism (on Xtensa, it's implemented with the WAITI instruction, literally "wait for interrupt").

It sounds like if you need this, maybe you shouldn't be calling k_cpu_idle() at all?

Reply:
@andyross  if we don't wait in k_cpu_idle(), pm_state_set will return to pm_system_suspend and go forward to pm_system_resume, then the core will be resumed.
````c
	k_sched_lock();
	pm_stats_start();
	/* Enter power state */
	pm_state_notify(true);
	atomic_set_bit(z_post_ops_required, id);
	pm_state_set(z_cpus_pm_state[id].state, z_cpus_pm_state[id].substate_id);  // power off
	pm_stats_stop();
        ....
	pm_stats_update(z_cpus_pm_state[id].state);
	pm_system_resume();  // resume
	k_sched_unlock();
````
Could you help to check https://github.com/zephyrproject-rtos/zephyr/pull/62522 ? Thanks!

Reply:
Right, you need to get to a WAITI, but the businesses of locking interrupts and calling the OS idle hook is what's weird.  `k_cpu_idle()` is an abstraction API provided by the OS to let tell the architecture layer to wait for the next interrupt, and that's not what your'e doing here[1].  You're writing SOC-specific code to trigger a hardware suspend in a particular way (in this case "disable interrupts and WAITI", I guess?).

Are you sure you don't want something like:

    __asm__ volatile("rsil %0, 0xf; waiti" : "=r"(dummy_ps));

instead?  The OS can't mess that up by misinterpreting you.

[1] At least one issue is that it does **not** always call WAITI!  I know there is (or at least was, once, not sure about the current status) a kconfig that would cause k_cpu_idle() to be a busy loop instead (for testing idle issues, mostly).  That obviously doesn't do what you want at all.

Reply:
@andyross Thanks for your comments! One issue that is we have such hw limitation, so we have to use it
````c
k_cpu_idle()
  arch_cpu_idle()
{
	/* Cribbed from SOF: workaround for a bug in some versions of
	 * the LX6 IP.  Preprocessor ugliness avoids the need to
	 * figure out how to get the compiler to unroll a loop.
	 */
	if (IS_ENABLED(CONFIG_XTENSA_WAITI_BUG)) {
#define NOP4 __asm__ volatile("nop; nop; nop; nop");
#define NOP32 NOP4 NOP4 NOP4 NOP4 NOP4 NOP4 NOP4 NOP4
#define NOP128() NOP32 NOP32 NOP32 NOP32
		NOP128();
#undef NOP128
#undef NOP32
#undef NOP4
		__asm__ volatile("isync; extw");
	}
...
}
````

Reply:
Ooh, digression: do you **actually** need this workaround anymore?  I'm the one that copied that out of the SOF code originally, where it's very old (APL-era, or maybe earlier) and not very well explained.  I never saw an errata note or anything explaining it (@jcmvbkbc might know?  I love that we have that address in the proverbial rollodex now!), and AFAIK we never hit the issue on any of the newer DSPs.

But if you do need that (and after submitting some docs explaining on exactly which hardware, etc...) maybe the right refactoring would be to put this in the arch layer as e.g. `xtensa_waiti_idle()` or something, and then calling that out of your SOC code.  Then you could use the same utility for both without abusing the kernel abstractions.

Reply:
On idea I had before closing the browser window: maybe the real issue is that all those nops are required **not** by the Xtensa core itself but by the Intel DSP power management hardware?  If that's the case (and obviously I have no idea) then the correct path is to add it back here at the SOC level only and then remove it from the Xtensa arch entirely.  It's really very expensive (384 bytes of extra code size, maybe 256 if nop will narrow in the assembler).

Reply:
> Ooh, digression: do you **actually** need this workaround anymore? I'm the one that copied that out of the SOF code originally, where it's very old (APL-era, or maybe earlier) and not very well explained. I never saw an errata note or anything explaining it (@jcmvbkbc might know? I love that we have that address in the proverbial rollodex now!), and AFAIK we never hit the issue on any of the newer DSPs.
> 
> But if you do need that (and after submitting some docs explaining on exactly which hardware, etc...) maybe the right refactoring would be to put this in the arch layer as e.g. `xtensa_waiti_idle()` or something, and then calling that out of your SOC code. Then you could use the same utility for both without abusing the kernel abstractions.

This workaround is for LX6 which is used by TGL. The CONFIG_XTENSA_WAITI_BUG is enabled for TGL. It is not for power management but for waiti requirement on intel platform.

yes, we need to rework the code for arch_cpu_idle, but it is not linked to this pr directly.  Can I submit another PR to address it ?  thanks!

Reply:
@andyross I think warrants more explanation. @RanderWang we should avoid talking about application stuff like SET_DX , Zephyr reviewers have no idea what this means. If we need to give application use examples (e.g. from SOF), we need to explain this better and use terms can be understood in context of Zephyr.

But basicly, we are trying to extend support for these older cAVS2.5 platforms in Zephyr. I think @andyross your point is correct. But, but, the thing is:
   1) SOF application uses Zephyr SOFT_OFF state in implementing app logic to shutdown certain DSP cores. SOF calls:

```
"!pm_state_force(id, &(struct pm_state_info){PM_STATE_SOFT_OFF, 0, 0}))"
````

2) to enable CONFIG_PM on cAVS2.5 in SOF, we need to implement at least the bare minimum of SOFT_OFF (what this PR does)

3) in this specific case, the core does not need to wake up anymore. in the only usage we have, the core will be powered down. we might later improve this, but for now, entering idle with no means to return is ok

So not sure @andyross if we really need to avoid k_cpu_idle() here. Would just a comment do explaining that we are disabling interrupt on purpose here...?

Reply:
@RanderWang Did you consider blocking interrupts on SOF side of the code? I think the problem here is a race with an IDC interrupt I guess...?

Reply:
To be clear, I'm not exactly a -1 on this.  I'm just pointing out that k_cpu_idle() is being abused here.  Architectures have a lot of latitude as to what happens there (c.f. the busy idle thing I mentioned earlier), and that makes relying on it when what you want is a hyper-specific hardware sequence ("disable interrupts, nop for 128 cycles, issue an instruction and memory barrier, and then waiti").

It's fragile, basically.  If you're writing hardware code you should rely on a hardware abstraction, which isn't what k_cpu_idle() is.  My suggestion was to add that to the arch layer ("arch_xtensa_core_idle()", something like that) and have k_cpu_idle() and this both use the same thing.

Reply:
Oh, and a link to an errata number added wouldn't hurt either.  Like I said I just copied this blindly, but it would be good to have a reference so integrators know which hardware actually needs this.

Reply:
And just to add: the point being there are a lot of LX6 devices out there in the market AFAICT SOF on Intel is the only platform that has stumbled on this.  Presumably all those other folks need it too, and it would be good for the OS arch layer to have a definitive answer for when to turn on the kconfig, or on which versions of the IP the bug is fixed, etc... (gratuitous ping of @jcmvbkbc again, heh).

Reply:
> @RanderWang Did you consider blocking interrupts on SOF side of the code? I think the problem here is a race with an IDC interrupt I guess...?

we power off the core after the pipeline is free and no more ipc message for this core, so no IDC interrupt for pipeline.  Other secondary cores can raise IDC interrupt to this core which happen in zephyr side.